libvast/builtins/commands/show.cpp

```diff
@@ -40,71 +144,69 @@ caf::message show_command(const invocation& inv, caf::actor_system& sys) {
+          if (show_concepts) {
+            auto concepts_definition
+              = to_definition(taxonomies.concepts, filter);
+            command_result->insert(command_result->end(),
+                                   concepts_definition.begin(),
+                                   concepts_definition.end());
           }
-        }
-        if (inv.full_name == "show" || inv.full_name == "show models") {
-          for (auto& [name, model] : taxonomies.models) {
-            auto definition = list{};
-            definition.reserve(model.definition.size());
-            for (auto& definition_entry : model.definition)
-              definition.push_back(std::move(definition_entry));
-            auto entry = record{
-              {"model",
-               record{
-                 {"name", std::move(name)},
-                 {"description", std::move(model.description)},
-                 {"definition", std::move(definition)},
-               }},
-            };
-            result.push_back(std::move(entry));
+          if (show_models) {
+            auto models_definition = to_definition(taxonomies.models, filter);
+            command_result->insert(command_result->end(),
+                                   models_definition.begin(),
+                                   models_definition.end());
           }
-        }
-        if (as_yaml) {
-          if (auto yaml = to_yaml(data{std::move(result)}))
-            fmt::print("{}\n", *yaml);
-          else
-            command_result = caf::make_message(std::move(yaml.error()));
-        } else {
-          if (auto json = to_json(data{std::move(result)}))
-            fmt::print("{}\n", *json);
-          else
-            command_result = caf::make_message(std::move(json.error()));
-        }
-      },
-      [&](caf::error& err) mutable {
-        command_result = caf::make_message(
-          caf::make_error(ec::unspecified, fmt::format("'show' failed to get "
+        },
+        [&](caf::error& err) mutable {
+          command_result = caf::make_error(ec::unspecified,
+                                           fmt::format("'show' failed to get "
                                                        "taxonomies from "
                                                        "catalog: {}",
-                                                       std::move(err))));
-      });
-  return command_result;
+                                                       std::move(err)));
+        });
+  }
+  if (show_schemas && !command_result.error()) {
+    auto catch_all_query = expression{predicate{
+      meta_extractor{meta_extractor::type},
+      relational_operator::not_equal,
+      data{""},
+    }};
+    auto query_context
+      = query_context::make_extract("show", self, std::move(catch_all_query));
+    query_context.id = uuid::random();
+    self
+      ->request(catalog, caf::infinite, atom::candidates_v,
+                std::move(query_context))
```

Comment:
Why exactly do we have to send a query when there is a handler to retrieve all known types?
This puts unnecessary stress on the catalog.

Reply:
This handler with this query is really quite cheap. The ask was to print all schemas for which we have data, and there is no other handler that can get me this information.

Reply:
I see that we don't make a full pass for negations, but this is not one. You should be able to turn the O(n) into O(1) by using `negation{expression{}}` instead.

Reply:
That's a neat idea. Never thought of that one before.